runtime/jsTest/src/kotlinx/serialization/DynamicParserTest.kt

```diff
@@ -168,4 +176,28 @@ class DynamicParserTest {
         val dyn1 = js("({data: 42})")
         assertEquals(NDWrapper(NotDefault(42)), deserializer.parse(dyn1))
     }
+
+    @Test
+    @Ignore // Currently broken!
```

Comment:
There's some problem with `pos` being incremented too late / the tag being read too early with maps. "type" is read from index 0 which then calls `decodeStringElement` which ends up calling `getByTag("type")` but because `pos` is at 0, not 1, the tag is returned directly rather than `obj[tag]`.

I couldn't figure out the problem, but figured this was good enough to land as-is because it at least allows array-based polymorphism to work.

Reply:
I have a feeling that it's more than an off-by-one error, because regular polymorphic serializer expects first a class discriminator in _primitive_ string, and then a start of object. Here, class discriminator is located directly inside object, so it likely won't work without additional special handling, similar to one that performed in JSON: https://github.com/kotlin/kotlinx.serialization/blob/95dc4312d8148c5b945bcb3961e2926feaeef5e2/runtime/commonMain/src/kotlinx/serialization/json/internal/Polymorphic.kt#L32 (i.e. basically intercepting `deserialize` call to AbstractPolymorphicSerializer and replacing it with custom logic of manipulating json tree).

Since it is some additional amount of work, you can leave this test as is with some more meaningful comment.

Reply:
The challenge with this form of polymorphism is that the type discriminator (if it is not merely the presence/absence of certain attributes is that it normally requires reading the entire json subtree into memory.

Thinking about how to "solve" this, it is certainly something that should be handled at format level, not at serializer level (it is json specific). You could use an extension of the json format that would have an additional configuration for polymorphism (possibly dependent on base type exactly like SerialModule). This configuration would be a function that would accept a `JsonElement` or `JsonObject` and would then return an actual value.

In the format it would hook into (`Encoder`|`Decoder`)`.beginStructure` where it can inspect the passed in serial descriptor for the kind (Polymorphic) and override the default behaviour (even if the serializer wants to write the type in a certain way, the format is free to ignore that - my XML format does not require type discriminators - the tag names are already).

Reply:
In Moshi we have a special peeking reader which you can fork off from the main reader. This allows you to peek and consume JSON for things like locating the "type" key for polymorphic deserialization. But instead of moving forward in the actual underlying data where going backwards is challening, it merely buffers the data in the original reader as you progress. Once you're done with the peeked reader, you throw it away and resume consuming the original reader (which is still in the same original position, but with N bytes buffered where N was the number of bytes read by the peeked reader). It means JSON in the form `{"type":"a","value":"b"}` and `{"value":"b","type":a"}` can both be correctly supported without having to convert the whole object to a tree-like model.

I'm not familiar enough with the codebase here to know whether such a thing could be applied, but it's an extremely useful concept for cases like this where you need to arbitrarily look ahead.

> Since it is some additional amount of work, you can leave this test as is with some more meaningful comment.

Thanks, will do.